### PR TITLE
CSHARP-2820: Add SDAM test for debouncing topology description updates.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
Evergreen:  https://evergreen.mongodb.com/version/5dd45a573627e00a0c1682db

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/Misc/EndPointHelper.cs
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Misc/EndPointHelper.cs:132]: before this method was not used anywhere

File: tests/MongoDB.Driver.Core.Tests/Core/Helpers/MockClusterableServerFactory.cs
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Core.Tests/Core/Helpers/MockClusterableServerFactory.cs]: Since `result.Monitor` is only interface, I created `ShouldDescriptionChangedEventBePublished` as a static method to be able to call it here

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs
Comment (jyemin) [inline src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs]: The specification states:

> Server description for each server MUST be updated (replaced) every heartbeat. However, new description MUST NOT cause the SDAM flow to be executed if the new description is equal, as defined in this section, to the previous description.

I'm not positive, but this change looks like it trades one bug for another.  With this change, the SDAM flow will, correctly, not be executed (and no events will be published either), but it will also fail to update the server description within the cluster description.  It's important that this update occurs, so that server selection sees the latest roundTripTime value (which is not included in the equality comparison).

It's definitely possible I'm misunderstanding the code, but this is what I suspect.

**Review by DmitryLukyanov — COMMENTED** [review]
Evergreen: https://evergreen.mongodb.com/version/5dd711712fbabe414107e320
Ready for rereview.

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs]: thank you for clarification. In the current version I left only logic to supress `ServerDescriptionChangedEvent` and `ClusterDescriptionChangedEvent` events

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs]: Nit 1: I believe the `object` qualifiers in this function are redundant
Nit 2: Would it be clearer to say `WireVersionRange.Equals`, `TagSet.Equals`, and `ElectionId.Equals`?
Question: What does `// me` stand for?

File: tests/MongoDB.Driver.Core.Tests/Core/Helpers/MockClusterableServerFactory.cs
Comment (vincentkam) [inline tests/MongoDB.Driver.Core.Tests/Core/Helpers/MockClusterableServerFactory.cs]: Nit: EOF should have a newline.

**Review by vincentkam — APPROVED** [review]
Two nits, one minor question, otherwise LGTM! 😸

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs]: >Nit 1: I believe the object qualifiers in this function are redundant

yes, but we do it in this way in a lot of other places. So I just repeated it.

Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs]: >Nit 2: Would it be clearer to say WireVersionRange.Equals, TagSet.Equals, and ElectionId.Equals?

it would not be the same logic. `object.Equals` protects variables from `NullReferenceException`.

>Question: What does // me stand for?

I guess it was discussed here: https://mongodb.slack.com/archives/GQ27C9N7J/p1575490389037500
Let me know if you still have any questions.

File: tests/MongoDB.Driver.Core.Tests/Core/Helpers/MockClusterableServerFactory.cs
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Helpers/MockClusterableServerFactory.cs]: done :)

File: src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs]: Thanks! And oops you're right about the object.Equals providing protection from nulls, reference, thanks for the explanation! :)

File: src/MongoDB.Driver.Core/Core/Servers/ServerDescriptionChangedEventArgs.cs
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Servers/ServerDescriptionChangedEventArgs.cs]: This seems like the wrong place to be determining this.

Seems like the event should be published and it should be up to the handler (**not** the event raiser) to decide whether the SDAM state should be changed or not.

In other words, the event should only describe *what* happened, not *how* to process it (which is what `ShouldSdamDescriptionChangedEventsBePublished` is doing).

I think this class should be left alone and the decision about how to process the event should be made at a higher level in the event handler.

File: src/MongoDB.Driver.Core/Core/Misc/EndPointHelper.cs
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Misc/EndPointHelper.cs]: The else is redundant.

**Review by rstam — CHANGES_REQUESTED** [review]
See comments regarding at which level the decision about how to process an event should be made.

Also, I don't think your current implementation satisfies this spec requirement:

Note: Server description for each server MUST be updated (replaced) every heartbeat. However, new description MUST NOT cause the SDAM flow to be executed if the new description is equal, as defined in this section, to the previous description.

**Comment by vincentkam** [comment]
Just making a note that we also have https://jira.mongodb.org/browse/SPEC-1517 to consider, which grew out of @jyemin 's concerns: should we hold off on merging?

**Comment by vincentkam** [comment]
Sorry =\ I think I have to withdraw my LGTM until we figure out whether this PR should be merged given the issue that @jyemin pointed out in https://jira.mongodb.org/browse/SPEC-1517

**Comment by DmitryLukyanov** [comment]
Not ready for review

**Comment by DmitryLukyanov** [comment]
Sorry I had merging issues that's why there is a number of `force-push`s. Now it's solved.
EG:  https://evergreen.mongodb.com/version/5e69119961837d55aa2c81f4
Ready for review.
NOTE: I didn't make splitting comments in this PR into the two different as it was discussed on the standup. I will do it after the whole review.
NOTE2: I will create a new ticket with SDAM bug description later today.

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/MultiServerClusterTests.cs
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Core.Tests/Core/Clusters/MultiServerClusterTests.cs]: Now, since we remove a server when it is not valid for the current cluster type => we see two new remove events

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/MultiServerClusterTests.cs
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Core.Tests/Core/Clusters/MultiServerClusterTests.cs:685]: the second event is suppressed based on the original ticket logic

File: tests/MongoDB.Driver.Core.Tests/Core/Helpers/MockClusterableServerFactory.cs
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Core.Tests/Core/Helpers/MockClusterableServerFactory.cs]: if the server has been removed, we possible will need to recalculate his description

File: tests/MongoDB.Driver.Core.Tests/Specifications/server-discovery-and-monitoring/tests/rs/repeated.yml
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Core.Tests/Specifications/server-discovery-and-monitoring/tests/rs/repeated.yml]: this file contains the test which led us to the last commit in this PR with this change: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/83/files#diff-346021f0b586b815ec28dc10028a0a5aR321-R322

**Comment by DmitryLukyanov** [comment]
The latest force-push was about reverting one line in test (no real changes in the code). Still ready for review.

**Comment by DmitryLukyanov** [comment]
EG (after rebasing on master): https://evergreen.mongodb.com/version/5e6924c2d1fe07709cb1226c

**Review by rstam — CHANGES_REQUESTED** [review]
Please re-request my review after you have rebased on CSHARP-3012.

**Comment by DmitryLukyanov** [comment]
ready for review

File: src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs:227]: I think the `Event` should *always* be raised.

It is up to the receiver to decide what to do (or not do).

File: src/MongoDB.Driver.Core/Core/Servers/Server.cs
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Servers/Server.cs:244]: I think the `Event` should *always* be raised.

It is up to the receiver to decide what to do (or not do).

File: src/MongoDB.Driver.Core/Core/Servers/ServerDescription.cs
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Servers/ServerDescription.cs]: Perhaps name the variable `sdamEquals`.

Or better yet... remove the variable and return directly?

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs:227]: See comment in earlier review on this same topic:

https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/83#discussion_r354090870

File: src/MongoDB.Driver.Core/Core/Servers/ServerDescription.cs
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Servers/ServerDescription.cs]: done

File: src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs:227]: I'm not sure how it can look like. At least if we pass an event into EventCapturer, then this event will be asserted in the test runner. So, if I remove this new part in the above condition, this test https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/83/files?file-filters%5B%5D=.cs&file-filters%5B%5D=.json#diff-97b7222883d33cefce025d2b22e41bb5R2 will fail

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs:227]: I see now what confused me...

We raise this event TWICE in this method, once using our own event mechanism and once using C# events.

Your change is ONLY suppressing the FIRST of the two events, so we are in fact ALWAYS raising the SECOND event, which matches what I expected in my first comment in this conversation.

File: src/MongoDB.Driver.Core/Core/Servers/Server.cs
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Servers/Server.cs:244]: I see now what confused me...

We raise this event TWICE in this method, once using our own event mechanism and once using C# events.

Your change is ONLY suppressing the FIRST of the two events, so we are in fact ALWAYS raising the SECOND event, which matches what I expected in my first comment in this conversation.

**Review by rstam — APPROVED** [review]
LGTM

File: src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs:326]: Is there anywhere in the spec that specifies this logic should happen? I'm guessing it's the table here? https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring-monitoring.rst#events

Also, is there a spec test for this logic? I would imagine that the `Monitoring a standalone connection - suppress update events for equal server descriptions"` results in a SingleServerCluster being created?

File: src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs:225]: Do we need a better name for this variable? Perhaps something related to SDAM Monitoring?
`shouldSdamMonitoringClusterDescriptionChangedEventBePublished`? A bit long, but it might help prevent the confusion that happened below?

File: src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs:229]: Similarly, should we rename this to `_sdamMonitoringDescriptionChangdEventHandler`, to make it absolutely clear that this is not the sdam flow event, but the sdam monitoring flow event?

**Review by vincentkam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs:326]: >Is there anywhere in the spec that specifies this logic should happen?

do you mean the spec requirements about where we need to trigger such events?

>Also, is there a spec test for this logic? 

no, but it sounds like a good idea to add it, I will work on it

File: src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs:229]: I don't really like such long name `_sdamMonitoringDescriptionChangdEventHandler`, but I agree that it makes sense to make a note(somehow) about the actual logic here. I will think about it

File: src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs:225]: I will think about it

**Review by DmitryLukyanov — COMMENTED** [review]
Thanks @vincentkam for notes about a missed test. I will work on it after streamable ismaster.

File: src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs:225]: @vincentkam , I'm not sure whether `monitoring` is exactly the sdam part that has been skipped via this `shouldClusterDescriptionChangedEventBePublished` property. In other words, can we say that `SDAM` (server discovery and monitoring) includes two parts: 
1. Server discovery => everything in sdam flow except user-level(diagnostic) events?
2. Monitoring => only is about user-level(diagnostic) events?

if not so, I would just add a comment similar to: it was decided to suppress only sdam events and do not touch sdam flow itself.

Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs:225]: I may be wrong, but my current understanding of the flow is that any event we raise that is backed by an `Action` is related to SDAM Monitoring, and that anything that is backed by an `EventHandler` is related to the actual SDAM flow/logic. In other words, I _think_ that `shouldClusterDescritionChangedEventBePublished` is merely supressing the SDAM Monitoring event.

Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs:225]: Notably, the [spec ticket](https://github.com/mongodb/specifications/commit/44efef18ea13cc1c201104052984ccd0b16da462) associated with the drivers ticket seems to only contain changes to the monitoring tests?

**Review by vincentkam — CHANGES_REQUESTED** [review]

**Review by vincentkam — APPROVED** [review]
Upon further reflection, I think the name of the variables can be rightfully described as a nit :) Therefore, I'll leave the decision on whether to change the names or to add a clarifying comment up to you Dima! LGTM :)  ᓚᘏᗢ  😸

File: src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs:225]: Since the PR is approved, let's continue discussing this case in the scope of new ticket which will be created for https://jira.mongodb.org/browse/DRIVERS-1249 :)

**Comment by DmitryLukyanov** [comment]
EG after rebasing: https://evergreen.mongodb.com/version/5ece54d69ccd4e37670e681e

**Pull request closed without merging**
